### PR TITLE
state.change listener with deep hash check

### PR DESCRIPTION
**Pull request by aliabid94**
state.change event listeners now do "deep checks", by deep hashing the old value. This makes @gr.render logic much cleaner. See demo/todo_list changes to see how much simpler the code is now because state values can be edited directly to induce a change.

Closes: https://github.com/gradio-app/gradio/issues/8436

### CODE REVIEW

**Comment by pngwn** [comment]
How expensive is this?

**Comment by aliabid94** [comment]
> How expensive is this?

Pretty cheap I would think. It runs a hash function for every element in the nested object, and only gets run if there's a state change listener involved. In regular python the same hash function is called whenever you add any element to a dictionary or a set, so hashing is already quite optimized.

File: gradio/utils.py [hunk 1]
```diff
@@ -1402,3 +1403,25 @@ def connect_heartbeat(config: dict[str, Any], blocks) -> bool:
...
+                for k, v in sorted(obj.items(), key=lambda x: hash(x[0]))
+            ]
+        )
+    elif isinstance(obj, (list, tuple)):
+        items = tuple(deep_hash(x) for x in obj)
+    elif isinstance(obj, set):
+        items = tuple(deep_hash(x) for x in sorted(obj, key=hash))
+    else:
+        items = str(id(obj)).encode("utf-8")
+    hasher.update(repr(items).encode("utf-8"))
```
Comment (abidlabs) [inline gradio/utils.py:1426]: Is there a reason you use `repr` here instead of `str`?

**Comment by abidlabs** [comment]
It might be worth adding a note on the `State.change()` event listener docstring that this could get expensive if the `State` object is a sequence with a large number of elements

**Review by abidlabs — APPROVED** [review]
Tested and works well, thanks @aliabid94!

File: gradio/utils.py (hunk 1 above)
Comment (aliabid94) [reply gradio/utils.py:1426]: its what chatgpt suggested lol

Comment (abidlabs) [reply gradio/utils.py:1426]: repr is more unique so that makes sense ✔️

**Comment by aliabid94** [comment]
> It might be worth adding a note on the State.change() event listener docstring that this could get expensive if the State object is a sequence with a large number of elements

I don't think it's possible to directly add docs to the event of a single component. Regardless will update docs and guides on gr.State in a follow up PR.

**Pull request merged**